docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+</aside>
+
+- Provides an interface `getPayloadLog` to build an upkeep payload for a particular log on demand by giving a trigger as in input.
+It gets only the required log from the log buffer or reads it from log poller if not found in buffer. This is used by the recovery flow
+
+#### Log Buffer
+
+A circular/ring buffer of blocks and their corresponding logs, that act as a cache for logs. 
+It is used to store the last `lookbackBlocks*3` blocks, where each block holds a list of that block's logs. 
+Logs are marked as seen when they are returned by the buffer, to avoid working with logs that have already been seen.
+
+It is used by the log provider to provide unknown logs to the node, and by by the log recoverer to identify known logs during recovery flow.
+
+
+### Log Recoverer
+
+The log recoverer is responsible to ensure that no logs are missed.
+It does that by running a background process for re-scanning of logs and putting the ones we missed into the recovery flow (without checkBlockNum/Hash).
+
+Logs will be considered as missed if they are older than `latestBlock - lookbackBlocks` and has not been performed or successfully checked already (eligible).
+
+While the provider is scanning latest logs, the recoverer is scanning older logs in ascending order, up to `latestBlock - lookbackBlocks`, newer blocks will be under the provider's lookback window.
+
+**Recoverer scanning process**
+
+- The recoverer maintains a `lastRePollBlock` for each upkeep, .i.e. the last block it scanned for that upkeep.
+- Every second, the recoverer will scan logs for a subset of `n=10` upkeeps, where `n/2` upkeeps are randomly chosen and `n/2` upkeeps are chosen by the oldest `lastRePollBlock`.
+- It will start scanning from `lastRePollBlock` on each iteration, and update the block number when it finishes scanning.
+- Logs that are older than 24hr are ignored, therefore `lastRePollBlock` starts at `latestBlock - (24hr block)` in case it was not populated before.
+
+#### Upkeep States
+
+The upkeeps states are used to track the status of log upkeeps (eligible, performed) across the system, to avoid redundant work by the recoverer. Enables to select by (upkeepID, logIdentifier) is used as a key to store the state of a log upkeep.
+
+The state is updated by the coordinator when the upkeep is performed or after positive check (eligible).
+
+The states will be persisted to so the latest state to be restored when the node starts up.
+
+### Plugin
+
+The plugin is performing the following tasks upon OCR3 procedures:
+
+#### Observation
+
+Observation starts with a processing of the previous outcome:
+- Remove agreed upon finalized results from result store
+- For `acceptedSamples` (bound to a trigger)
+    - Remove from metadata store
+    - Enqueue (trigger, upkeepID) into conditional observer
+- For `acceptedRecoveryLogs` (bound to a trigger)
+    - Remove from metadata store
+    - Enqueue (trigger, upkeepID) into recovery observer
```

Comment:
Both of these processes also use the latest coordinated block from the previous outcome. If there isn't a previous coordinated block in the recent outcome history, the results are ignored.